MH-13381, Use Organization Identifier in Workflows

lkiesow opened the pull request:
The maximum size of the serialized workflow instances stored in the
database is huge, causing an unnecessary performance overhead for the
frequent serialization and deserialization to/from XML. One cause for
this is the large organization data stored multiple times in the XML.

This patch drops the use of the serialized organization which is not
frequently being used anyway and replaces it by referencing the
organization via its unique identifier.

The old organization is still kept as an optional field, ensuring
backward compatibility to old serialized workflows. This means that no
migration is required.

There are more references to organizations in the XML caused by the user
data stored in the XML, but they will be dealt with separately.

lkiesow committed:
MH-13381, Use Organization Identifier in Workflows

The maximum size of the serialized workflow instances stored in the
database is huge, causing an unnecessary performance overhead for the
frequent serialization and deserialization to/from XML. One cause for
this is the large organization data stored multiple times in the XML.

This patch drops the use of the serialized organization which is not
frequently being used anyway and replaces it by referencing the
organization via its unique identifier.

The old organization is still kept as an optional field, ensuring
backward compatibility to old serialized workflows. This means that no
migration is required.

There are more references to organizations in the XML caused by the user
data stored in the XML, but they will be dealt with separately.

smarquard commented:
Confirmed by looking at payload in oc_job that the workflow serialized XML now has:

    <sec:organizationid>mh_default_org</sec:organizationid>

rather than

    <sec:organization id="mh_default_org"> ... (etc) ... </sec:organization>

Still to do some basic regression testing, but looks good.

A similar change for the sec:creator user info will be a big win as well.

staubesv commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstanceImpl.java:
```diff
@@ -202,8 +210,15 @@ public User getCreator() {
   @Override
-  public Organization getOrganization() {
-    return organization;
+  public String getOrganization() {
```
getOrganizationId() would be more consistent with other similar situations in the codecase

staubesv commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstanceImpl.java:
```diff
@@ -217,14 +232,11 @@ public void setCreator(User creator) {
-      this.organization = null;
-    else
-      this.organization = JaxbOrganization.fromOrganization(organization);
+  public void setOrganization(final String organizationId) {
```
setOrganizationId() would be more consistent with other similar situations in the codecase

staubesv commented on modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java:
```diff
@@ -2387,7 +2387,13 @@ public void repopulate(final String indexName) throws Exception {
-        Organization organization = instance.getOrganization();
+        Organization organization = null;
+        try {
+          organization = organizationDirectoryService.getOrganization(instance.getOrganization());
```
... getOrganization() expects the ID of the organization. While instance.getOrganization() now indeed does return the ID, instace.getOrganizationId() would make this more obvious.

staubesv commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstanceImpl.java:
```diff
@@ -88,6 +93,9 @@
   @XmlElement(name = "organization", namespace = "http://org.opencastproject.security")
```
That likely can be removed...

staubesv reviewed (dismissed):
I've found a minor issue:

- The setter/getter method set-/getOrganization() should be renamed to set-/getOrganizationId()

[Update] Just found the serialized workflow instance in the DB....

staubesv commented on modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowInstanceImpl.java:
```diff
@@ -88,6 +93,9 @@
   @XmlElement(name = "organization", namespace = "http://org.opencastproject.security")
   private JaxbOrganization organization;
 
+  @XmlElement(name = "organizationid", namespace = "http://org.opencastproject.security")
```
Maybe organization_id or organizationId for sake of readability?

lkiesow replied:
I changed it to `organization-id` to be consistent with other XML tags we use like for example `time-in-queue`

lkiesow committed:
MH-13381, Organization Id Method Naming

This patch renames the methods returning the organization identifier to
clearly distinguish them from those methods returning the organization
object.
